encodings/fastlanes/src/for/compute.rs

```diff
@@ -65,21 +65,15 @@ impl FilterFn<FoRArray> for FoREncoding {
 
 impl ScalarAtFn<FoRArray> for FoREncoding {
     fn scalar_at(&self, array: &FoRArray, index: usize) -> VortexResult<Scalar> {
-        let encoded_pvalue = scalar_at(array.encoded(), index)?
-            .into_value()
-            .as_pvalue()
-            .vortex_expect("Encoded scalar must be primitive")
-            .map(|p| p.reinterpret_cast(array.ptype()));
-        let reference = array
-            .reference()
-            .as_pvalue()
-            .vortex_expect("Reference scalar must be primitive")
-            .vortex_expect("Reference scalar cannot be null");
+        let encoded_pvalue = scalar_at(array.encoded(), index)?.reinterpret_cast(array.ptype());
+        let encoded_pvalue = encoded_pvalue.as_primitive();
+        let reference = array.owned_reference_scalar();
+        let reference = reference.as_primitive();
 
         Ok(match_each_integer_ptype!(array.ptype(), |$P| {
             encoded_pvalue
-                .map(|v| v.as_primitive::<$P>().vortex_unwrap())
-                .map(|v| v.checked_shl(array.shift() as u32).unwrap_or_default().wrapping_add(reference.as_primitive::<$P>().vortex_unwrap()))
+                .typed_value::<$P>()
+                .map(|v| v.checked_shl(array.shift() as u32).unwrap_or_default().wrapping_add(reference.typed_value::<$P>().vortex_expect("reference must be non-null")))
```

Comment:
Looks like typed_value returns None iff the PValue is None, i.e., the value is null. If `array.ptype()` is not the type of the scalar I think we actually panic because of the `vortex_expect` in `typed_value`.